File: modelstore-frontend/src/app/shared/interfaces/product.ts
import { Guid } from 'guid-typescript';

export class Product {
  id: Guid;
  name: string;
  brand: string;
  slug: string;
  rating: number | null;
  price: number;
  stock: number;
  categories: number[];
  description: string;
  imageUrl?: string;
  promotion?: number;
  categoriesText: string;

  constructor(
    id: Guid,
    name: string,
    brand: string,
    slug: string,
    rating: number | null,
    price: number,
    stock: number,
    categories: number[],
    description: string,
    categoriesText: string,
    imageUrl?: string,
    promotion?: number
  ) {
    this.id = id;
    this.name = name;
    this.brand = brand;
    this.slug = slug;
    this.rating = rating;
    this.price = price;
    this.stock = stock;
    this.categories = categories;
    this.description = description;
    this.categoriesText = categoriesText;
    this.imageUrl = imageUrl;
    this.promotion = promotion;
  }
}

File: ModelStore.Application/Models/Product.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace ModelStore.Application.Models
{
    public partial class Product
    {
        public Guid Id { get; init; }

        public string Name { get; set; }

        public string Brand { get; set; }

        public string Slug => GenetateSlug();

        public float? Rating { get; set; }

        public decimal Price { get; set; }

        public int Stock { get; set; }

        public List<int> Categories { get; init; } = new();

        public string Description { get; set; }

        public int? Promotion { get; set; }

        private string GenetateSlug()
        {
            var sluggedName = SlugRegex().Replace(Name, string.Empty)
                .ToLower().Replace(" ", "-");

            var sluggedBrand = Regex.Replace(Brand, "[^0-9A-Za-z _-]", string.Empty)
                .ToLower().Replace(" ", "-");

            return $"{sluggedName}-{sluggedBrand}";
        }

        [GeneratedRegex("[^0-9A-Za-z _-]", RegexOptions.NonBacktracking, 5)]
        private static partial Regex SlugRegex();
    }
}